### PR TITLE
feat(vrl): Support DataDog grok parser - parsing DD grok patterns

### PR DESCRIPTION
**Pull request by vladimir-dd**
This is the first PR related to parsing DD grok patterns to break down #7705 in smaller chunks.

Signed-off-by: Vladimir Zhuk <vladimir.zhuk@datadoghq.com>

Here is an example of DD grok patterns(can be seen as an extension of "normal" grok patterns):
    %{number:duration:scale(1000000000)} 
    %{regex("(check\\s)?\\w+"):function.name}

You can write parsing rules with the `%{MATCHER:EXTRACT:FILTER}` syntax:

- Matcher: A rule (possibly a reference to another token rule) that describes what to expect (number, word, notSpace, etc.)
- Extract (optional): An identifier representing the capture destination for the piece of text matched by the Matcher.
- Filter (optional): A post-processor of the match to transform it.

### CODE REVIEW

File: lib/parsing/Cargo.toml [hunk 1]
```diff
@@ -0,0 +1,21 @@
+[package]
+name = "parsing"
```
Comment (StephenWakely) [inline lib/parsing/Cargo.toml:2]: I'm a bit confused by this crate. It doesn't seem to contain any parsing code. 

Is this moving the `Value` type  out of `vector-core` or creating a new `Value` type?

Comment (vladimir-dd) [reply lib/parsing/Cargo.toml:2]: Ah, right. This package will contain some shared parsing logic(ruby hash, xml etc). For example, if there is a vrl function which does the same thing as one of the grok filters we put the shared logic there.

**Review by StephenWakely — APPROVED** [review]
Nice, looks good.

File: lib/parsing/Cargo.toml (hunk 1 above)
Comment (leebenson) [reply lib/parsing/Cargo.toml:2]: Agreed, I think "parsing" feels like too generic a name (especially hosted directly under `lib`).

I'd probably keep grok and VRL code separate, even if there's some redundancy/overlap there, unless there's a strong reason for this to be coupled.

**Review by leebenson — APPROVED** [review]
Looks great @vladimir-dd 🎉 

Just echoed one comment re: the "parsing" lib. I'd probably keep this logic within the `lib/datadog` namespace for now (or `lib/vrl`, if more applicable), as it's not clear looking at it as a standalone library what it's intended to do.

If there's a good reason to keep it as a shared lib (i.e. if you know iterative refactoring of #7705 is likely to lean on it), then moving / renaming to fit the intended use-case and adding some comments would provide some helpful context.

**Comment by vladimir-dd** [comment]
> Looks great @vladimir-dd 🎉
> 
> Just echoed one comment re: the "parsing" lib. I'd probably keep this logic within the `lib/datadog` namespace for now (or `lib/vrl`, if more applicable), as it's not clear looking at it as a standalone library what it's intended to do.
> 
> If there's a good reason to keep it as a shared lib (i.e. if you know iterative refactoring of #7705 is likely to lean on it), then moving / renaming to fit the intended use-case and adding some comments would provide some helpful context.

So this shared logic is supposed to be used by both grok and vrl libs, therefore I wanted to keep it somewhere outside. @leebenson given this, do you think I could keep it as a standalone library after adding more comments? don't have an idea of a better name though, maybe you have any suggestions? it's basically a library which contains different parsing methods(xml, ruby-hash, query string etc)

**Comment by leebenson** [comment]
> ... given this, do you think I could keep it as a standalone library after adding more comments? don't have an idea of a better name though, maybe you have any suggestions? it's basically a library which contains different parsing methods(xml, ruby-hash, query string etc)

The `Value` that you're creating here is the same as the struct in `vector-core`, right?

On first glance, it feels like there's two parts here:

1. A shared `Value`, which would vector-core, vrl and grok would depend on. I think calling that one `lib/value` would be fine.
2. Shared parsing methods. I think these could be housed in VRL, if Datadog libs are already relying on VRL libs and these directly serve VRL functions.

I'd keep implementation of the `Value` in grok / VRL separately to keep those concerns distinct.

**Comment by vladimir-dd** [comment]
@leebenson  as we discussed I've removed this custom `Value` type - it is basically a subset of vrl `Value` so we can use it instead. In the follow-up PRs I'll add parsing functions to the vrl lib.

**Review by JeanMertz — COMMENTED** [review]
Apparently I started reviewing this PR, but never finished it. Apologies. I submitted the review anyway, since it might be useful for a follow-up PR.

**Pull request merged**